### ISSUE TITLE
Export a prettier-compatible configuration

### ISSUE THREAD

**Issue by fregante** [stylelint-stylistic/stylelint-stylistic#27]
### What is the problem you're trying to solve?

Prettier and Stylistic work on similar parts of formatting, but Stylistic has additional rules that are still interesting to Prettier users.

Before v16, you could just apply https://github.com/prettier/stylelint-config-prettier after any shared `stylelint` config and call it a day:

```json
	"stylelint": {
		"extends": [
			"stylelint-config-xo",
			"stylelint-config-prettier"
		],
```

### What solution would you like to see?

I think this plugin could export something similar to `stylelint-config-prettier`, which means it would `null` any rules that conflict with prettier

**Comment by firefoxic** [comment]
@fregante, hi!
I'm not sure that I understood correctly what result you want to achieve. But I will assume that you are missing a config that provides those rule settings that were removed with the `stylelint@16` release. If so, I can suggest using [`@stylistic/stylelint-config`](https://github.com/stylelint-stylistic/stylelint-config#readme).

Let me know if this does not solve your problem.

**Comment by fregante** [comment]
Refer to https://github.com/prettier/stylelint-config-prettier

In short, there should be a config that **overrides and disables* anything that causes conflict with Prettier. The repo above can explain it better if it's not clear.

An example exported config:

```json
		"rules": {
			"@stylistic/declaration-colon-newline-after": null,
			"@stylistic/indentation": null,
			"@stylistic/selector-combinator-space-before": null,
			"@stylistic/selector-descendant-combinator-no-non-space": null,
			"@stylistic/value-list-comma-newline-after": null,
            … and anything that conflicts with prettier
```

**Comment by firefoxic** [comment]
I realized that my suggestion above was the exact opposite. You don't need `@stylistic/stylelint-confif`. Use `@stylistic/stylelint-plugin`.

The point is that unlike a config, a plugin doesn't do anything by itself. It just provides rules that can be enabled and customized to suit the needs of the project. In the case of `prettier/stylelint-config-prettier`, you had to override a huge number of rules because it is a pre-configured config. But using `@stylistic/stylelint-plugin` you can simply use only the rules you need in your config. It works the same way with the rules of `stylelint` itself - just don't use what you don't need.

**Comment by MorevM** [comment]
@firefoxic 

Well, the use of `prettier` itself suggests that one would be more inclined to trust the plugin developer on what is good, rather than configuring rules oneself.

It's understandable - the configuration for all the rules looks something [like this](https://github.com/MorevM/stylelint-config/blob/master/configurations/css/plugins/stylistic.js) and not everyone is ready to put time and take responsibility for it.

I've seen many such configurations, and they usually look like this:
```js
export default {
  extends: [
    'stylelint-config-standard',
    'stylelint-config-recommended-scss',
    'stylelint-config-someone-i-believe-knows-better'
  ],
  rules: {
    // Granular DISABLING of rules I don't like, 
    // rather than ENABLING the ones I do like.
  }
}
```

While I too disapprove this approach (and the `prettier` usage as well), it has its place and is widely used, and it is useful for a plugin to provide such a configuration(s) like `@stylistic/stylelint-config-recommended` and `@stylistic/stylelint-config-prettier`

We can discuss defaults for the recommended (possibly standard) configuration, then I can do that.
Implementing an prettier-compatible configuration after that is a minor detail.

I do believe that introducing pre-built configurations can expand the plugin's audience, and indeed spread good practices more widely in general.

**Comment by firefoxic** [comment]
I will try to explain my position in detail.

### Firstly

Let's understand the history. It used to be that stylistic rules were inside Stylelint itself. And the `stylelint-config-standard` included many of them. And `stylelint-config-prettier` used to disable all those rules.

Now, stylistic rules have been removed not only from Stylelint, but also from the `stylelint-config-standard`. That's why `stylelint-config-prettier` is now deprecated — it's no longer needed, because `prettier` has nothing else to conflict with.

For clarity, I've tabulated all the rules that are disabled in `stylelint-config-prettier` and noted which ones are in `@stylistic/stylelint-plugin` and which ones are in `@stylistic/stylelint-config`:

| # | Presence of a rule in… | `@stylistic` plugin | `@stylistic` config |
|---|---|---|---|
| 1 | `at-rule-empty-line-before` | ❌ | ❌ |
| 2 | `at-rule-name-case` | ✔️ | ✔️ |
| 3 | `at-rule-name-newline-after` | ✔️ | ❌ |
| 4 | `at-rule-name-space-after` | ✔️ | ✔️ |
| 5 | `at-rule-semicolon-newline-after` | ✔️ | ❌ |
| 6 | `at-rule-semicolon-space-before` | ✔️ | ❌ |
| 7 | `block-closing-brace-empty-line-before` | ✔️ | ✔️ |
| 8 | `block-closing-brace-newline-after` | ✔️ | ✔️ |
| 9 | `block-closing-brace-newline-before` | ✔️ | ✔️ |
| 10 | `block-closing-brace-space-after` | ✔️ | ❌ |
| 11 | `block-closing-brace-space-before` | ✔️ | ✔️ |
| 12 | `block-opening-brace-newline-after` | ✔️ | ✔️ |
| 13 | `block-opening-brace-newline-before` | ✔️ | ❌ |
| 14 | `block-opening-brace-space-after` | ✔️ | ✔️ |
| 15 | `block-opening-brace-space-before` | ✔️ | ✔️ |
| 16 | `color-hex-case` | ✔️ | ✔️ |
| 17 | `declaration-bang-space-after` | ✔️ | ✔️ |
| 18 | `declaration-bang-space-before` | ✔️ | ✔️ |
| 19 | `declaration-block-semicolon-newline-after` | ✔️ | ✔️ |
| 20 | `declaration-block-semicolon-newline-before` | ✔️ | ❌ |
| 21 | `declaration-block-semicolon-space-after` | ✔️ | ✔️ |
| 22 | `declaration-block-semicolon-space-before` | ✔️ | ✔️ |
| 23 | `declaration-block-trailing-semicolon` | ✔️ | ✔️ |
| 24 | `declaration-colon-newline-after` | ✔️ | ✔️ |
| 25 | `declaration-colon-space-after` | ✔️ | ✔️ |
| 26 | `declaration-colon-space-before` | ✔️ | ✔️ |
| 27 | `function-comma-newline-after` | ✔️ | ✔️ |
| 28 | `function-comma-newline-before` | ✔️ | ❌ |
| 29 | `function-comma-space-after` | ✔️ | ✔️ |
| 30 | `function-comma-space-before` | ✔️ | ✔️ |
| 31 | `function-max-empty-lines` | ✔️ | ✔️ |
| 32 | `function-parentheses-newline-inside` | ✔️ | ✔️ |
| 33 | `function-parentheses-space-inside` | ✔️ | ✔️ |
| 34 | `indentation` | ✔️ | ✔️ |
| 35 | `linebreaks` | ✔️ | ❌ |
| 36 | `max-empty-lines` | ✔️ | ✔️ |
| 37 | `max-line-length` | ✔️ | ✔️ |
| 38 | `media-feature-colon-space-after` | ✔️ | ✔️ |
| 39 | `media-feature-colon-space-before` | ✔️ | ✔️ |
| 40 | `media-feature-name-case` | ✔️ | ✔️ |
| 41 | `media-feature-parentheses-space-inside` | ✔️ | ✔️ |
| 42 | `media-query-list-comma-newline-after` | ✔️ | ✔️ |
| 43 | `media-query-list-comma-newline-before` | ✔️ | ❌ |
| 44 | `media-query-list-comma-space-after` | ✔️ | ✔️ |
| 45 | `media-query-list-comma-space-before` | ✔️ | ✔️ |
| 46 | `no-empty-first-line` | ✔️ | ✔️ |
| 47 | `no-eol-whitespace` | ✔️ | ✔️ |
| 48 | `no-extra-semicolons` | ✔️ | ✔️ |
| 49 | `no-missing-end-of-source-newline` | ✔️ | ✔️ |
| 50 | `number-leading-zero` | ✔️ | ✔️ |
| 51 | `number-no-trailing-zeros` | ✔️ | ✔️ |
| 52 | `property-case` | ✔️ | ✔️ |
| 53 | `selector-attribute-brackets-space-inside` | ✔️ | ✔️ |
| 54 | `selector-attribute-operator-space-after` | ✔️ | ✔️ |
| 55 | `selector-attribute-operator-space-before` | ✔️ | ✔️ |
| 56 | `selector-attribute-quotes` | ❌ | ❌ |
| 57 | `selector-combinator-space-after` | ✔️ | ✔️ |
| 58 | `selector-combinator-space-before` | ✔️ | ✔️ |
| 59 | `selector-descendant-combinator-no-non-space` | ✔️ | ✔️ |
| 60 | `selector-list-comma-newline-after` | ✔️ | ✔️ |
| 61 | `selector-list-comma-newline-before` | ✔️ | ❌ |
| 62 | `selector-list-comma-space-after` | ✔️ | ❌ |
| 63 | `selector-list-comma-space-before` | ✔️ | ✔️ |
| 64 | `selector-max-empty-lines` | ✔️ | ✔️ |
| 65 | `selector-pseudo-class-case` | ✔️ | ✔️ |
| 66 | `selector-pseudo-class-parentheses-space-inside` | ✔️ | ✔️ |
| 67 | `selector-pseudo-element-case` | ✔️ | ✔️ |
| 68 | `string-quotes` | ✔️ | ✔️ |
| 69 | `unicode-bom` | ✔️ | ❌ |
| 70 | `unit-case` | ✔️ | ✔️ |
| 71 | `value-list-comma-newline-after` | ✔️ | ✔️ |
| 72 | `value-list-comma-newline-before` | ✔️ | ❌ |
| 73 | `value-list-comma-space-after` | ✔️ | ✔️ |
| 74 | `value-list-comma-space-before` | ✔️ | ✔️ |
| 75 | `value-list-max-empty-lines` | ✔️ | ✔️ |

Which can be seen from the table:

1. There are a total of 75 rules in `stylelint-config-prettier` (and all of them are only disabled).
2. Of these rules, `@stylistic/stylelint-plugin` has only 73 (and even less in `@stylistic/stylelint-config`).
3. The remaining 2 rules are still part of Stylelint itself (i.e. they have nothing to do with the `@stylistic` project).

### Conclusion

Previously, when using `stylelint-config-standard` together with `prettier` you had to disable stylistic rules with `stylelint-config-prettier`. Now, if you still want to use `stylelint-config-standard` with `prettier` for styles, just don't plug in `@stylistic/stylelint-plugin` or even `@stylistic/stylelint-config`. The most you may need to do is to disable those very 2 rules:

```json
{
	"extends": [
		"stylelint-config-standard"
	],
	"rules": {
		"at-rule-empty-line-before": null,
		"selector-attribute-quotes": null
	}
}
```

If you still think that disabling these two rules should be in some plugin, I'm fine with that, but that plugin definitely has nothing to do with the `@stylistic` project, because these rules are inside Stylelint itself.

**Comment by fregante** [comment]
`stylelint-config-standard` is just _one_ config. `stylelint-config-xo` includes and enables this plugin, so in order to be used with Prettier, some of its rules need to be disabled.

https://github.com/xojs/stylelint-config-xo/blob/dd178d358aad3064a9e4ac8347b669d415d7cced/index.js#L23

**Comment by MorevM** [comment]
@firefoxic 

It's the wrong counting mechanism.

**A couple of examples:**

`@stylistic/stylelint-config` [specifies](https://github.com/stylelint-stylistic/stylelint-config/blob/main/.stylelintrc.json#L33) that two spaces are used for indentation. 
`prettier` config [may specify](https://prettier.io/docs/en/options.html#tabs) that `tab` should be used for indentation. Or [another count of spaces](https://prettier.io/docs/en/options.html#tab-width).

`@stylistic/stylelint-config` [specifies](https://github.com/stylelint-stylistic/stylelint-config/blob/main/.stylelintrc.json#L64) that double quotes should be used. 
`prettier` config [may specify](https://prettier.io/docs/en/options.html#quotes) that single ones should be used.

So there is a conflict here. 
If I want to use `prettier` together with `@stylistic/stylelint-config` configuration, I need to disable them.

---

By the way: I didn't dive too deep (and I don't use `prettier` at all), but except for properties above, the `@stylistic/stylelint-config` configuration at first glance looks identical to what `prettier` does.

If I'm not mistaken, for `prettier` users it could be recommended to simply remove `stylelint-stylistic` from the `stylelint` configuration and nothing would change. 
However, `stylelint-stylistic` can provide additional value - take at least the [`named-grid-areas-alignment`](https://github.com/stylelint-stylistic/stylelint-stylistic/tree/main/lib/rules/named-grid-areas-alignment) rule I recently added (although it is not included in `@stylistic/stylelint-config`, but it should be). 
Prettier does not provide such option. 
Someday I'll deal with the new rules that are hanging in open issues - and it's unlikely to be ever included in `prettier`.
So I'm against recommending “just do not use it”.

Even if I'm right about confliction rules (and there are really only two), "just do not use it" solution is not documented anywhere and it makes perfect sense that people run into such issues during migration, package upgrades.

I extracted the count of `stylelint` downloads by version, here are the results:

Major Stylelint version | Weekly downloads
-------------------------|----------------------
16.x.x                            | 1 079 861 
15.x.x                            | 1 144 373
14.x.x                            | 1 220 383
13.x.x                            | 691 298

So every one of those two million, when migrating from version 15 to 16, will be wondering “OK, they removed stylistic rules, OK, there's a replacement in the form of a plugin, but what should I do about `prettier` compatibility?”.
It would be nice to provide them with an easy way to solve this problem while staying in a familiar environment, ideally just replacing one line with a different line.

---

Also, as noted above by @fregante, some people provide their configurations without knowing whether the person using it is using `prettier` or not - and it would be nice to be able to provide the same behavior. 

Of course, each author of such a configuration can make their own research, figure out which `stylelint-stylistic` rules will conflict with `prettier`, and disable those rules themselves. 
And then `prettier` will decide to change something and it will have to be done again for each shareable configuration... 
A single source of truth is needed here.

**Comment by fregante** [comment]
> So every one of those two million, when migrating from version 15 to 16, will be wondering “OK, they removed stylistic rules, OK, there's a replacement in the form of a plugin, but what should I do about prettier compatibility?”.

I like this one. 

*To migrate to Stylelint 16:*

- *Add Stylistic*
- *If present, replace prettier config with Stylistic/prettier-override*

**Comment by firefoxic** [comment]
Sorry for the delayed response; I needed time to consider and discuss a lot of things. However, this response will be final.

So. It was logical and correct to have a plugin in the organization that disables something that was interfering with prettier itself. But back then, there was no choice because the `stylelint-config-standard` included stylistic rules. Now, if you need to use prettier, you can simply avoid using any config or plugin with stylistic rules.

Your situation, where you are using both `prettier` and someone else's config that includes stylistic rules (and therefore isn't designed to work with `prettier`), is quite unusual and exceedingly rare. You wouldn't use two incompatible tools for the same task in one project, like `yarn` and `pnpm` for package management — clearly, you need to drop one of them. The same goes for `prettier` and `@stylistic/stylelint-plugin` — you need to drop one of them, at least for styles.

But having something that disables `@stylistic/stylelint-plugin` rules to make `prettier` work (let me remind you that **the complete non-acceptance of `prettier` was the reason for creating the `@stylistic/stylelint-plugin`**), is an even stranger situation. It's like having a bakery with a sign on the door saying, “Don't come in here, the croissants at the competitor around the corner are better” — absurd!

I also considered a migration guide. However, Stylelint's migration guide to version 15 is fully focused on `prettier`. And only as an alternative (not an addition), `@stylistic/stylelint-plugin` is suggested. So if we need a guide, it should be about migrating from `stylelint-config-standard` to `stylelint-config-standard` + `@stylistic/stylelint-config`, which is already available in the README of the config. So, there's nothing more to do here.

Therefore, I am closing this issue. But if you still need help configuring your setup (disabling interfering rules), let me know — I am ready to send a pull request to your project.

**Comment by fregante** [comment]
You missed one important part:

> Stylistic has additional rules that are still interesting to Prettier users.

I get that Stylistic tries to do what Prettier does, but that's not true for the opposite. Prettier lacks `color-hex-case`, `number-leading-zero`, and all the rules that **alter the code** and not just the spacing/line-breaks.

Admittedly, these rules are the minority but that doesn't mean that "you need to drop one of them" and that it's "absurd"

**Comment by firefoxic** [comment]
> Prettier lacks `color-hex-case`, `number-leading-zero`, and all the rules that **alter the code** and not just the spacing/line-breaks.

I agree with this. But it doesn't cancel the fact that conflict of `prettier` settings with `@stylelint/stylistic-plugin` rules occurs only in such rare cases as using someone else's config for `@stylelint/stylistic-plugin` together with `prettier`. And in such cases it is more logical to resolve the conflict (disable conflicting rules) inside the project itself (such as yours).

All other cases, including using `@stylelint/stylistic-plugin` without foreign configs and together with `prettier`, do not need separate npm-packages with config to disable rules. Because there is nothing to disable — rules that conflict with `prettier` are simply not included. Even more often `Prettier` is not present at all or is disabled for styles — then there can be no conflicts.

And I am still ready to help you with PR in your project 🙂
